gk/bpf.c

```diff
@@ -156,6 +201,12 @@ gk_bpf_encapsulate(struct gk_bpf_pkt_ctx *ctx, int priority)
 	if (unlikely(priority < 0 || priority > PRIORITY_MAX))
 		return -EINVAL;
 
+	if (priority == PRIORITY_GRANTED && ctx->direct_if_possible) {
+		return update_pkt_priority(frame->pkt,
+			frame->fe->grantor_fib->u.grantor.gt_addr.proto,
```

Comment:
GK has already parsed `frame->pkt`, but `gk_bpf_decide_pkt()` is not pushing `struct ipacket` down. Modify `gk_bpf_decide_pkt()`, so their caller pass `struct ipacket` instead of just the packet. This way, `update_pkt_priority()` won't have to parse the Ethernet header again, and we don't have to "guess" the type of the header looking at `frame->fe->grantor_fib->u.grantor.gt_addr.proto`, but just checking the field `flow.proto`.